### PR TITLE
[Color system] Add color factory and build modern tokens

### PR DESCRIPTION
**Pull request by danrosenthal**
Merging ~the next branch into master, which includes several changes after the v2.7.0 release, as well as the latest~ changes incorporating the color factory into tokens, and using TypeScript in the build for modern tokens.

Release candidate on UNPKG: https://unpkg.com/browse/@shopify/polaris-tokens@2.8.0-rc.0/dist-modern/

This will all be released as v2.8.0, since these changes are not breaking.

### CODE REVIEW

**Comment by danrosenthal** [comment]
@kaelig, should we be merging f471512 and 68772b3 into master along with this, or should those remain unmerged in the `next` branch?

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,18 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 <!-- ## [Unreleased] -->
```
Comment (kaelig) [inline CHANGELOG.md]: Nit pick

```suggestion
## [Unreleased]
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,18 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
 
+- Added color factory and built modern tokens ([#105](https://github.com/Shopify/polaris-tokens/pull/105))
+  - Added surface disabled variant and updated other variant configs ([#101](https://github.com/Shopify/polaris-tokens/pull/101))
+  - Added TypeScript build for modern tokens, and shifted directory structures to differentiate between legacy and modern tokens ([#97](https://github.com/Shopify/polaris-tokens/pull/97))
+  - Updated variant names and descriptions ([#96](https://github.com/Shopify/polaris-tokens/pull/96))
+  - Added decorative icon variants ([#94](https://github.com/Shopify/polaris-tokens/pull/94))
+  - Built changes from previous release, and added textOnInteractive variant ([#93](https://github.com/Shopify/polaris-tokens/pull/93))
+  - Fixed an issue where legacy themes caused the color factory to throw ([#92](https://github.com/Shopify/polaris-tokens/pull/92))
+  - Update color variants for consistency with changes in Polaris React ([#91](https://github.com/Shopify/polaris-tokens/pull/91))
```
Comment (kaelig) [inline CHANGELOG.md]:
```suggestion
  - Updated color variants for consistency with changes in Polaris React ([#91](https://github.com/Shopify/polaris-tokens/pull/91))
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -7,6 +7,18 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+  - Marked the config as optional and the colors as partial ([dd3d8fc](https://github.com/Shopify/polaris-tokens/commit/dd3d8fc05572fb03e764a85a0519bbd3dde11855))
+  - Actually exporting the types would be helpful ([4998856](https://github.com/Shopify/polaris-tokens/commit/49988564d7048c73e53cd85eab89f5ff49d0e28b))
```
Comment (kaelig) [inline CHANGELOG.md]: Not sure this message is useful for the reader?

```suggestion
  - Actually exporting the types would be helpful ([4998856](https://github.com/Shopify/polaris-tokens/commit/49988564d7048c73e53cd85eab89f5ff49d0e28b))
```

File: docs/index.css [hunk 4]
```diff
@@ -121,7 +120,7 @@ thead th:first-child {
 }
 
 code {
-  font-family: Monaco, Consolas, 'Lucida Console', monospace;
+  font-family: Monaco, Consolas, "Lucida Console", monospace;
```
Comment (kaelig) [inline docs/index.css]: I thought the convention was single quotes – can you check if there is a Sass or Prettier configuration issue?

File: CHANGELOG.md (hunk 3 above)
Comment (danrosenthal) [reply CHANGELOG.md]:
```suggestion
```

File: docs/index.css (hunk 4 above)
Comment (danrosenthal) [reply docs/index.css]: @BPScott updated the prettier config to not format these files, so not an issue.

Ben, what was your rationale there?

Comment (BPScott) [reply docs/index.css]: This was deliberatly changed - The docs folder is [no longer formatted with prettier](https://github.com/Shopify/polaris-tokens/pull/105/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R17) (but [I do use sass's expanded style](https://github.com/Shopify/polaris-tokens/pull/105/files#diff-b9e12334e9eafd8341a6107dd98510c9R253) to maintain most of the readable formatting  :)

I split the build into "legacy", "modern" and "docs" steps, this makes it easy to do just the modern build in shipit (to generate the `dist-modern` folder), and also change the docs build to something new (probably a proper gatsby app) in the near future. In order to retain the old behaviour I would have had to either:

- have two steps "build-legacy:format" (to format the dist folder with prettier) and "docs:format" (to format the css in the docs folder with prettier) to that run after each of the build-legacy and docs steps
- A single step that runs after both "build-legacy" and "docs" - which makes makes the of steps a bit more convoluted and less self-contained.

I figured seeing as this file is generated content, and is going away imminently it is easier to do create a build-legacy:format step that formats just the dist folder, and skip formatting the docs folder css with prettier, as sass already does a pretty solid job. Incidentally I've just realised by formatting this file after sass does its thing we risk breaking the css.map file that the sass build generates (as that maps locations in the not-prettiered css file, not the version that's been ran through prettier)

**Pull request merged**